src/utilities/apiImpl.ts

```diff
@@ -0,0 +1,51 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ */
+
+import * as sarifApi from "./../api/sarifViewerApi";
+import * as vscode from "vscode";
+import { LogReader } from "../logReader";
+import { SVDiagnosticCollection } from "../svDiagnosticCollection";
+import { openSarifFile } from "../extension";
+
+export class ApiImpl implements sarifApi.Api {
+    /**
+     * Constructs an instance of the API given out by the extension.
+     * @param logReader An instance of the log reader to use for parsing log files.
+     * @param diagnosticCollection The diagnostic collection that holds the results.
+     */
+    public constructor(private readonly logReader: LogReader, private readonly diagnosticCollection: SVDiagnosticCollection) {
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public async openLogFiles(sarifFiles: vscode.Uri[], openLogFileArguments?: sarifApi.OpenLogFileArguments): Promise<void> {
+        for (const sarifUri of sarifFiles) {
+            await openSarifFile(sarifUri, this.logReader, this.diagnosticCollection, {
+                closeOriginalFileOnUpgrade: true,
+                openInTextEditor: openLogFileArguments?.openInTextEditor === undefined ? true : openLogFileArguments.openInTextEditor,
+                promptUserForUpgrade: openLogFileArguments?.promptUserForUpgrade === undefined ? true : openLogFileArguments.promptUserForUpgrade
```

Comment:
`openLogFileArguments?.promptUserForUpgrade ?? true` #pattern

Reply:
I need to investigate how openLogFileArguments?.promptUserForUpgrade ?? true does when promptUserForUpgrade is undefined vs false

---
In reply to: [409120203](https://github.com/microsoft/sarif-vscode-extension/pull/248#discussion_r409120203) [](ancestors = 409120203)

Reply:
works as you demonstrated. Changing.

---
In reply to: [409143005](https://github.com/microsoft/sarif-vscode-extension/pull/248#discussion_r409143005) [](ancestors = 409143005,409120203)